userland/libtock/ble.c

```diff
@@ -1,43 +1,148 @@
+/*
+ * BLE setup functions
+ */
+
 #include "ble.h"
+#include "tock.h"
+#include <stdbool.h>
+#include <stdint.h>
 #include <stdio.h>
 #include <string.h>
 
-int ble_adv_set_txpower(BLE_TX_Power_t power) {
-  return command(DRIVER_RADIO, BLE_CFG_TX_POWER, power);
+#define MAX_SIZE 31
+
+
+/*******************************************************************************
+ *   INTERNAL BLE HELPER FUNCTION Prototypes
+ *
+ *   s_   - static (file scope)
+ ******************************************************************************/
+static int s_ble_cfg_addr (uint8_t *addr, uint8_t addr_len);
+static int s_ble_cfg_flags (uint8_t flags);
+static int s_ble_cfg_adv_itv (uint16_t advertising_itv_ms);
+static int s_ble_cfg_gap_data (GapAdvertisementData_t header, uint8_t *data, uint8_t len);
+
+
+/*******************************************************************************
+ *   USER-SPACE BLE API
+ ******************************************************************************/
+
+int ble_initialize(uint16_t advertising_itv_ms, bool discoverable, uint8_t *addr,
```

Comment:
I think a good question is whether we want to allow users-pace to set the BLE address, or whether that should be in the kernel (or potentially in a separate driver that eventually is only accessible by privileged processes or something).

Consider the situation where we have two processes sending advertisements. We can interleave different advertising packets from each process with the same address no problem. We can also, in principle, send different advertising packets with _different_ addresses from each process, but it becomes more complicated if they the device is connectable. Specifically, I'm not sure it will be as easy to have the device actually connect on two different addresses (though might be an interesting piece of engineering to attempt).

There are various ways we can imagine breaking ties, for example only actually responding to connection handshakes one address at a time, but it's worth considering what we actually _want_ to do.

Regardless it's probably also worth allowing the process to not choose an address (perhaps by passing `NULL` and `0` to `addr` and `addr_len`, respectively), and then defaulting to either an address a different process has already set, or something reasonable the kernel decides on (e.g. an address based on the hardware MAC address or something).

Reply:
I think the busy flag in the capsule should at least protect against "races" in changing advertising address for. 

However it may limit concurrency which also is valid question to consider as well.

Reply:
I didn't really think about it before, but it's probably a good idea for the kernel to set the BLE address of the device and for the application to have no input on it. However, this isn't how the nRF SDK works, so if we made that change, it would be a difference between serialized and native applications.

The exception would be if we really could virtualize everything about BLE, changing addresses with each advertisement and accepting connections to specific applications, but that sounds pretty tough.

(edit: I posted this somewhere else, but it really belongs here)

Reply:
Ok, but it sounds like it is best to let kernel decide the address for now then?

Should be pretty easy to use the TRNG driver to generate 6 random bytes if you want that?

But should all functions be completely re-written because it is still possible have interleaving between configuring advertisement_interval, configure different data (device_name, uuids etc) in the packet.

But after ```_start``` the busy flag will be set and block the entire thing as it is right now

I think you @alevy and @brghena has to decide this because I don't know how to approach this.

Reply:
I think for now having the kernel control the address is best.

From the perspective of _this_ PR, just having passed into the driver as a configuration variable in the board setup seems fine. The point is it's not up to the process. There are several ways we can set the address in the kernel (e.g. there is also a 64-bit unique device identifier in the FICR controller for the NRF52; randomness could work too). But that's for a different PR, it's not about the interface.

As far as other functions in the userland interface are concerned, I don't think they are problematic in the same way. I think at the moment (i.e. given the current state of the driver in the kernel) interleaving won't work as we expect, but we know how to make it work. In particular, you'll just store the data for each process's configured advertising packet in a grant and, in the kernel, switch between which one you're sending to interleave them. So, for example, if I have two processes, one with a 50ms advertising interval, and another with a 100ms advertising interval, my actual on-air packets might look something like this:

```
Time         Packet
====         ======
0      --->  "process 1 packet"

25 ms  --->  "process 2 packet"

50 ms  --->  "process 1 packet"

75 ms  --->  no_data

100 ms --->  "process 1 packet"

125 ms --->  "process 2 packet"

150 ms --->  "process 1 packet"
```

Again, it's fine if the kernel driver doesn't do that just yet, as long as the interface supports it.

As for scanning, I think basically all incoming scanned packets should be delivered to all processes that have subscribed to scans. In the future we could maybe add functions that allow processes to configure filters for which packets they actually want to receive (and I can imagine implementing this in a combination of hardware and software in the kernel).

Reply:
This is definitely what I was thinking too Amit.